jskit/encoders/utils/models.py

```diff
@@ -224,11 +226,14 @@ def forward(self, context_input_ids=None, context_input_masks=None,
                     return candidate_vec
         if labels is not None and mode == "train":
             candidate_vec = candidate_vec.squeeze(1)
-            dot_product = torch.matmul(
-                context_vec, candidate_vec.t())  # [bs, bs]
-            mask = torch.eye(context_input_ids.size(0)).to(
-                context_input_ids.device)
-            loss = F.log_softmax(dot_product, dim=-1) * mask
+            if self.loss_type == 'dot-prod':
+                dot_product = torch.sum(context_vec * candidate_vec, dim=-1)  # row-wise dot product
+                loss = F.log_softmax(dot_product, dim=-1)
+            else:  # `cos-sim`
+                loss_fn = torch.nn.CosineEmbeddingLoss()
+                batch_size = candidate_vec.size(0)
+                # TODO: support negative pairs
```

Comment:
So this doesn't work with negative examples?

Reply:
https://github.com/Jaseci-Labs/jaseci/pull/147/commits/cf7c5afeb8828d9a20cbdbc56ab769e6a78d88cf#diff-eef0e0493716527b9006d448f729d8c4ff5f4a52988f05999bf95907d0e0e134R236

Here we assign only positive labels as the target. With respect to loss we can have it work with negative examples by just matching the labels from the batch. However, supporting negative examples requires more changes in the entire forward pass. 

So no it doesn't work with negative examples now.